### ISSUE TITLE
Rename the string.bytes iterator

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#13663]
We are about to make `bytes` a keyword to add the bytes type to the language. This'll require us to do something for [string.bytes() iterator](https://chapel-lang.org/docs/builtins/String.html#String.string.bytes) which iterates a string byte-by-byte.

Currently, in my branch where I have `bytes` as a keyword, I have changed `string.bytes` to `string.iterBytes`. I am not particularly attached to the idea. The PR: #13579

**Comment by e-kayrakli** [comment]
Copying @bradcray's [comment](https://github.com/chapel-lang/chapel/pull/13579#issuecomment-518827954):

> Given that it's a breaking change, I think that the renaming of string.bytes warrants some design discussion if it hasn't occurred already and I missed it... (and that it should take place on its own issue advertised by email).

>Personally, I'm not wild about the name string.iterBytes() because it doesn't quite read like English to me (for i in s.iterBytes()...) (and if I were choosing a noun for it, I might choose bytesIter()). Brainstorming:

    string.getBytes()
    string.byteIter()
    string.eachByte()

> I'm not crazy about the following suggestion, but... we could also consider having the parser support .bytes as a special method/member name and sugaring it to something else (as .domain is currently) or supporting overloads to it... The downside to doing so is that it suggests either having the compiler not do it for non-string types or making it a user-facing feature to create such a method/iterator. (equivalently, I've been inclined to either retire our .domain queries or making them user-facing, where I have a slight preference for retiring if we can agree on a replacement name. See issue #10939 for related discussion.

> Even if we went with the "just rename it" strategy, to help with the transition for existing codes (assuming .bytes() existed in 1.19?) we may even want to have the parser recognize the pattern and give a hint as to what the new name is...

**Comment by e-kayrakli** [comment]
In case we take the rename route: Among those 3 suggested by Brad, I don't like `getBytes` which sounds to me like something that would return a `bytes` object from the `string`. I don't have any concerns about `byteIter` or `eachByte`

**Comment by dmk42** [comment]
It'a an unfortunate situation.  If we do a rename, we should make that family of names consistent.
```chapel
string.byte()
string.bytes()
string.codepoint()
string.codepoints()
```
I don't like this much, but I dislike it less than the other options so far:
We could append the word element or elements.
```chapel
string.byteElement()
string.byteElements()
string.codepointElement()
string.codepointElements()
```

**Comment by dmk42** [comment]
If there is a rename, I guess my second choice would be this.
```chapel
string.byte()
string.byteIter()
string.codepoint()
string.codepointIter()
```

**Comment by mppf** [comment]
Actually at the moment I think we shouldn't rename it at all. We can make the parser allow methods named `bytes` `string` or `domain` without causing confusion IMO. In this particular case, I think the potential source of confusion is that `string.bytes()` is an iterator yielding the bytes but to convert a string to bytes one would probably use a cast.

**Comment by e-kayrakli** [comment]
I prefer all those and other primitive type names to be reserved in the language. It reduces flexibility while writing code but prevents confusion while reading it IMHO.
 
That being said, it seems that I don't have strong support in this. I can look into supporting `bytes` as method name. And then we can have both `string.bytes()` and `bytes.bytes()`

**Comment by e-kayrakli** [comment]
After some offline discussions, current consensus is to have `bytes` as a special keyword so that when it is used as parenful method call, it'll call the receiver's `chpl_bytes()`. Other occurrences of `bytes` will be interpreted as the `bytes` type.

This is similar to current behavior of the keyword `domain` (called function is `_dom` in that case)

Therefore, the string implementation will be renamed to `chpl_bytes` but there will be **no change to the string interface**. The changes are implemented in #13579 . Closing this issue.
